### PR TITLE
Add assertions for and auto-checking of Bugsnag-Integrity header

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

Adds `the Bugsnag-Integrity header is valid` as a step to validate the `Bugsnag-Integrity` header. This should be a valid SHA-1 or simple digest.

The validation logic is covered by new unit tests. This has also been run against the Android CI here: https://github.com/bugsnag/bugsnag-android/pull/987

### CODE REVIEW

File: lib/features/support/header.rb
Comment (twometresteve) [inline lib/features/support/header.rb:12]: Computing the digest after Maze Runner has parsed JSON text into a Ruby hash, and this has generated it back into JSON, isn't going to work.  Any differences in whitespace will lead to a difference in the digest and it needs to be computed "at source".

Perhaps if you can action the other comments, then assign the ticket to me and I'll put that in though.

File: lib/features/support/header.rb
Comment (fractalwrench) [reply lib/features/support/header.rb:14]: Removed logging of body

File: lib/features/support/header.rb
Comment (fractalwrench) [reply lib/features/support/header.rb]: Updated to check the digest type.

**Comment by fractalwrench** [comment]
@twometresteve addressed the simpler review comments, assigning the rest over to you 🎉

**Comment by twometresteve** [comment]
@imjoehaines - I performed the initial review here for @fractalwrench, but subsequently added some additional changes for the automatic checking of the`Bugsnag-Integrity` header.  If present, it must now be valid.

To test this I built the test fixture locally against the in-development Android notifier that now provides the header (hence migrating to the latest notifier version) and ran the tests with DEBUG on.  I was able to inspect the logs to verify that the headers were present and that the computed digests matched.

Note that some status checks are currently outstanding.  I have changed the Github branch protection rules after refactoring the pipeline tests on #164 .  It will come out in the wash once #164 is merged and this PR rebased.

File: CHANGELOG.md
Comment (fractalwrench) [inline CHANGELOG.md:5]:
```suggestion
- Add assertions for automatic verification of Bugsnag-Integrity headers
```

Comment (twometresteve) [reply CHANGELOG.md:5]: Ah, what I really meant was `Add assertions for, and automatic verification of, Bugsnag-Integrity headers`(with commas), because there are both opt-in Cucumber steps and the no-option, check every request I get feature that have been added.  I'll reword to make that clearer.

File: lib/features/support/header.rb
Comment (imjoehaines) [inline lib/features/support/header.rb:29]: This should use [`bytesize`](https://ruby-doc.org/core-2.7.2/String.html#method-i-bytesize), to handle multi-byte characters

```suggestion
  computed_digest = "simple #{body.to_s.bytesize}"
```

File: lib/features/support/servlet.rb
Comment (imjoehaines) [inline lib/features/support/servlet.rb:82]: This also needs to be `bytesize`

```suggestion
    simple = request.body.bytesize
```

File: lib/features/support/servlet.rb
Comment (imjoehaines) [inline lib/features/support/servlet.rb:74]: I wonder if we should have a flag that can be set in an after configuration hook, which would tell Maze Runner that a missing integrity header is an error (or even whether it should check the header at all)

The implicit checking is nice, but if it silently does nothing when the header isn't present then it feels a bit dangerous

e.g.

```ruby
AfterConfiguration do
  MazeRunner.config.check_integrity_header = true
end
```

```diff
+    return unless MazeRunner.config.check_integrity_header
+
     header = request['Bugsnag-Integrity']
-    return if header.nil?
+    raise 'No Bugsnag-Integrity header found' if header.nil?
```

**Review by imjoehaines — CHANGES_REQUESTED** [review]

File: lib/features/support/servlet.rb
Comment (twometresteve) [reply lib/features/support/servlet.rb:74]: I'm not against that, but if we did that then I'd prefer it got passed in as a command line option.  The `config` construction is more a by-product of that and more for internal MR use that for clients to interact with.

Also, in terms of checking that a header is there once implemented by the notifier, I saw that as where the notifier should have its own scenarios to explicitly check for it (as Android has).

Comment (twometresteve) [reply lib/features/support/servlet.rb:74]: ...that is, the intention of the implicit checking was an additional safeguard on top of what notifier should be checking for themselves, as opposed to it being our only safety net.

File: lib/features/support/header.rb
Comment (imjoehaines) [reply lib/features/support/header.rb:19]: > Computing the digest after Maze Runner has parsed JSON text into a Ruby hash, and this has generated it back into JSON, isn't going to work. Any differences in whitespace will lead to a difference in the digest and it needs to be computed "at source".

I think we still need to do this before merging

File: lib/features/support/servlet.rb
Comment (imjoehaines) [reply lib/features/support/servlet.rb:74]: Yeah I get it, but it's a bit of a shame to have to add assertions to the notifier to know that it's being checked. This definitely doesn't block merging, but I think it'd be nice to opt in to a stricter mode eventually

About config vs command line options; I'd prefer leaning towards configuring MR in ruby if it's something that is universal to all MR tests, e.g. if we added a command line option to the Ruby pipeline we'd need to edit 71 different steps — not hard to do, but it's easy to miss some or forget to add it to new steps in the future

Comment (twometresteve) [reply lib/features/support/servlet.rb:74]: OK, I've raised a fresh internal ticket (PLAT-5419) to look at that.  Good point about accessing config directly at the language level if it works for a given notifier, I just wouldn't want to force other platforms down that route.  So we can add an option and the entry in `MazeRunner.config` will happen naturally.

File: lib/features/support/header.rb
Comment (twometresteve) [reply lib/features/support/header.rb]: Well spotted, that almost slipped through the net.  I've made some changes to the steps and unit tests, but will rerun the Android test that relies on it as well to ensure that it's what is required.

**Review by imjoehaines — APPROVED** [review]

**Pull request merged**
